ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/BasicAction.java

```diff
@@ -1498,7 +1498,11 @@ protected synchronized int End (boolean reportHeuristics)
             }
             else
             {
-                if (prepare(reportHeuristics) == TwoPhaseOutcome.PREPARE_NOTOK) {
+                int prepareStatus = prepare(reportHeuristics);
+
+                if (prepareStatus == TwoPhaseOutcome.PREPARE_NOTOK
+                        || (prepareStatus == TwoPhaseOutcome.PREPARE_ONE_PHASE_COMMITTED
```

Comment:
You are looking at an outdated diff (sorry). I changed prepare to return
the correct status, as in:

`return actionStatus == ActionStatus.ABORTED ?
TwoPhaseOutcome.ONE_PHASE_ERROR :
TwoPhaseOutcome.PREPARE_ONE_PHASE_COMMITTED;`

On Thu, Dec 17, 2015 at 10:54 AM, Tom Jenkinson <notifications@github.com>
wrote:

> In
> ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/BasicAction.java
> <https://github.com/jbosstm/narayana/pull/964#discussion_r47892593>:
>
> > @@ -1498,7 +1498,11 @@ protected synchronized int End (boolean reportHeuristics)
> >              }
> >              else
> >              {
> > -                if (prepare(reportHeuristics) == TwoPhaseOutcome.PREPARE_NOTOK) {
> > +                int prepareStatus = prepare(reportHeuristics);
> > +
> > +                if (prepareStatus == TwoPhaseOutcome.PREPARE_NOTOK
> > +                        || (prepareStatus == TwoPhaseOutcome.PREPARE_ONE_PHASE_COMMITTED
>
> One option is to change onePhasePrepare API to return the right status
> https://github.com/jbosstm/narayana/blob/master/ArjunaCore/arjuna/classes/com/arjuna/ats/arjuna/coordinator/BasicAction.java#L2322.
> That would be a public API change so not advisible for this release but
> might be something to consider for future.
>
> How about returning TwoPhaseOutcome.PREPARE_NOTOK if the 1PC fails
> (determined by actionStatus in the place you are currently proposing)? We
> lose resolution as to the cause of the failure but it minimizes the
> changeset and the new return code would not actually be used anyway except
> in your changeset.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jbosstm/narayana/pull/964/files#r47892593>.
>



-- 
Michael Musgrove
Transactions Team
e: mmusgrov@redhat.com
t: +44 191 243 0870

Registered in England and Wales under Company Registration No. 03798903
Directors: Michael Cunningham (US), Paul Hickey (Ireland), Matt Parson
(US), Charles Peters (US)

Michael Cunningham (US), Charles Peters (US), Matt Parson (US), Michael
O'Neill(Ireland)